src/hooks/useRawSearchParams.ts

```diff
@@ -0,0 +1,44 @@
+import { useMemo } from 'react';
+
+type UpdateSearchParamsOptions = {
+  method?: 'replace' | 'push';
+};
+
+type UpdateSearchParams = (
+  newSearchParams: URLSearchParams,
+  options?: UpdateSearchParamsOptions,
+) => void;
+
+type RawSearchParams = [URLSearchParams, UpdateSearchParams];
+
+/**
+ * This hook provides an `updateSearchParams` function that can update the
+ * search parameters _without_ creating a re-render on the router level.
+ */
+const useRawSearchParams = (): RawSearchParams => {
+  // Memoize the current search parameters to avoid unnecessary recalculations
+  const searchParams = useMemo(() => {
+    return new URLSearchParams(window.location.search);
+  }, [window.location.search]);
+
+  // This function updates the search params without causing a re-render
+  // and provides flexibility to use either pushState or replaceState.
+  const updateSearchParams: UpdateSearchParams = (
+    newSearchParams,
+    options = { method: 'replace' },
+  ) => {
```

Comment:
In terms of API, what about having updateSearchParams take a record instead of a SearchParams? The record wouldn't be the full search param but how we'd like to change it. That would move the complexity from the user to the hook.

* if a property has a primary value => use "set"
* if a property has an array value => use "set" for the first one, "append" for the other ones
* if a property has a null or undefined value => use "delete"

This would be a departure from what react router's `useSearchParams` does (because that one always takes the full version, either a Record, a Array, a URLSearchParams or a string, see [type declaration in their doc](https://reactrouter.com/en/main/hooks/use-search-params)), but I think this fits better our use cases where we want to change only one part, at the price of some flexibility.

What do you think?

Reply:
We've discussed it in Matrix a bit, and decided that pros and cons are not so clear for one over another. So let's continue with this approach for now and then we can revisit later if we really want.